### PR TITLE
support dora finetune in mlx-examples/llms/mlx_lm

### PR DESCRIPTION
**Pull request by alexC-nonsense4k**
support dora finetune in mlx_lm.
dora  shows a better finetune effect than lora and with a even smaller rank.
I add a new arg dora:bool  in lora.py in mlx_lm. By setting it as true, the dora layers will be applied to model during finetune.

### CODE REVIEW

File: llms/mlx_lm/tuner/utils.py [hunk 1]
```diff
@@ -31,6 +31,117 @@ def build_schedule(schedule_config: Dict):
     else:
         return bound_schedule_fn
 
+def linear_to_dora_layers(
+    model: nn.Module,
+    num_dora_layers: int,
+    config: Dict,
+):
```
Comment (awni) [inline llms/mlx_lm/tuner/utils.py:38]: Instead of copying the function `linear_to_lora_layers` I would just pass the bool `use_dora` to `linear_to_lora_layers`.

File: llms/mlx_lm/tuner/utils.py [hunk 2]
```diff
@@ -31,6 +31,117 @@ def build_schedule(schedule_config: Dict):
...
+        keys = set(["norm_attn_norm.attn.Wqkv", "ffn.router.layer"])
+    else:
+        raise ValueError(f"Lora does not support {model.model_type}")
+
+    for l in model.layers[num_layers - num_dora_layers :]:
+        dora_layers = [(k, to_dora(m)) for k, m in l.named_modules() if k in keys]
+        l.update_modules(tree_unflatten(dora_layers))
+
+
+def apply_dora_layers(model: nn.Module, adapter_path: str) -> nn.Module:
```
Comment (awni) [inline llms/mlx_lm/tuner/utils.py:105]: Also don't duplicate this function, just use the bool that gets saved in the config.

File: llms/mlx_lm/tuner/utils.py [hunk 3]
```diff
@@ -31,6 +31,117 @@ def build_schedule(schedule_config: Dict):
...
+        raise FileNotFoundError(f"The adapter path does not exist: {adapter_path}")
+    with open(adapter_path / "adapter_config.json", "r") as fid:
+        config = types.SimpleNamespace(**json.load(fid))
+    linear_to_dora_layers(model, config.lora_layers, config.lora_parameters)
+    model.load_weights(str(adapter_path / "adapters.safetensors"), strict=False)
+    return model
+
+
+
+def remove_dora_layers(model: nn.Module) -> nn.Module:
```
Comment (awni) [inline llms/mlx_lm/tuner/utils.py:127]: Same comment. Just remove both LoRA and DoRA layers. (Also I don't think we use this function anymore 🤔 , so could also just remove it).

File: llms/mlx_lm/lora.py [hunk 4]
```diff
@@ -140,6 +140,7 @@ def build_parser():
         help="Use gradient checkpointing to reduce memory use.",
     )
     parser.add_argument("--seed", type=int, default=0, help="The PRNG seed")
+    parser.add_argument("--dora",type=bool,default=False,help= "Use DORA to finetune")
```
Comment (awni) [inline llms/mlx_lm/lora.py:143]: Add it to `CONFIG_DEFAULTS` above.

File: llms/mlx_lm/tuner/dora.py [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+
+        # Low rank lora weights
+        scale = 1 / math.sqrt(input_dims)
+        self.lora_a = mx.random.uniform(
+            low=-scale,
+            high=scale,
+            shape=(input_dims, r),
+        )
+        self.lora_b = mx.zeros(shape=(r, output_dims))
+        self.m = mx.array(self.linear.weight.square().sum(axis=0,keepdims=True).sqrt())
```
Comment (awni) [inline llms/mlx_lm/tuner/dora.py:107]: Should this not be the second axis? Also do this like:

```python
self.m = mx.linalg.norm(self.linear.weight, axis=1)
```

File: llms/mlx_lm/tuner/dora.py [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+    def __call__(self,x) :
+        dtype=self.linear.weight.dtype
+        lb = (self.scale * self.lora_b.T).astype(dtype)
+        la = self.lora_a.T.astype(dtype)
+        adapted=self.linear.weight+lb.astype(dtype) @ la
+        norm= adapted.square().sum(axis=0,keepdims=True).sqrt()
+        y = self.linear(x)
+        z = (self.dropout(x) @ self.lora_a) @ self.lora_b
+        res=y+(self.scale*z).astype(x.dtype)
```
Comment (awni) [inline llms/mlx_lm/tuner/dora.py:116]: For this to be correct, apply the weight first `y = x @ self.linear.weight.T` then at the end add the bias if it is there.


```python
if "bias" in self.linear:
  res = res + self.linear.bias
```

File: llms/mlx_lm/tuner/dora.py [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+            shape=(input_dims, r),
+        )
+        self.lora_b = mx.zeros(shape=(r, output_dims))
+        self.m = mx.array(self.linear.weight.square().sum(axis=0,keepdims=True).sqrt())
+    def __call__(self,x) :
+        dtype=self.linear.weight.dtype
+        lb = (self.scale * self.lora_b.T).astype(dtype)
+        la = self.lora_a.T.astype(dtype)
+        adapted=self.linear.weight+lb.astype(dtype) @ la
+        norm= adapted.square().sum(axis=0,keepdims=True).sqrt()
```
Comment (awni) [inline llms/mlx_lm/tuner/dora.py:113]:
```suggestion
        norm= mx.linalg.norm(adapted, axis=1)
```

Same question here, should it not be the second axis? (E.g. `axis=1`)

File: llms/mlx_lm/tuner/dora.py [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+        r: int = 8,
+        alpha: float = 16,
+        dropout: float = 0.0,
+        scale: float = 10.0,
+    ):
+        # TODO remove when input_dims and output_dims are attributes
+        # on linear and quantized linear
+        output_dims, input_dims = linear.weight.shape
+        if isinstance(linear, nn.QuantizedLinear):
+            input_dims *= 32 // linear.bits
```
Comment (awni) [inline llms/mlx_lm/tuner/dora.py:22]:
```suggestion
        if isinstance(linear, nn.QuantizedLinear):
            raise ValueError("DoRALinear does not yet support quantized weights.")
```

File: llms/mlx_lm/tuner/dora.py [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+        dora_lin = DoRALinear(
+            linear=linear,
+            input_dims=input_dims,
+            output_dims=output_dims,
+            r=r,
+            alpha=alpha,
+            dropout=dropout,
+            scale=scale,
+        )
+        #dora_lin.linear = linear
```
Comment (awni) [inline llms/mlx_lm/tuner/dora.py:32]: ?

File: llms/mlx_lm/tuner/dora.py [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+        return dora_lin
+    
+    
+    def to_linear(self, de_quantize: bool = False):
+        linear = self.linear
+        bias = "bias" in linear
+        weight = linear.weight
+        m=self.m
+
+        is_quantized = isinstance(linear, nn.QuantizedLinear)
```
Comment (awni) [inline llms/mlx_lm/tuner/dora.py:42]: Remove

File: llms/mlx_lm/tuner/dora.py [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+
+        if is_quantized:
+            dtype = mx.float16
+            weight = mx.dequantize(
+                weight,
+                linear.scales,
+                linear.biases,
+                linear.group_size,
+                linear.bits,
+            )
```
Comment (awni) [inline llms/mlx_lm/tuner/dora.py:55]: Remove

File: llms/mlx_lm/tuner/dora.py [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+        output_dims, input_dims = weight.shape
+        fused_linear = nn.Linear(input_dims, output_dims, bias=bias)
+
+        lora_b = (self.scale * self.lora_b.T).astype(dtype)
+        lora_a = self.lora_a.T.astype(dtype)
+        fused_linear.weight = weight + lora_b @ lora_a
+
+        fused_norm=fused_linear.weight.square().sum(axis=0,keepdims=True).sqrt()
```
Comment (awni) [inline llms/mlx_lm/tuner/dora.py:63]: `mx.linalg.norm`

File: llms/mlx_lm/tuner/dora.py [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+
+        fused_linear.weight=fused_linear.weight/fused_norm*m
```
Comment (awni) [inline llms/mlx_lm/tuner/dora.py:65]:
```suggestion
        fused_linear.weight =  (m / fused_norm) * fused_linear.weight
```

File: llms/mlx_lm/tuner/dora.py [hunk 14]
```diff
@@ -0,0 +1,119 @@
...
+
+        if bias:
+            fused_linear.bias = linear.bias
+        if is_quantized and not de_quantize:
+            fused_linear = nn.QuantizedLinear.from_linear(
+                fused_linear,
+                linear.group_size,
+                linear.bits,
+            )
```
Comment (awni) [inline llms/mlx_lm/tuner/dora.py:74]: Remove.

File: llms/mlx_lm/tuner/dora.py (hunk 7 above)
Comment (awni) [reply llms/mlx_lm/tuner/dora.py:113]: Also these should have a stop gradient on it I believe `mx.stop_gradient`

**Comment by alexC-nonsense4k** [comment]
@awni , I sincerely appreciate your thorough code review. I will address all the issues promptly and implement your valuable suggestions. They will undoubtedly assist me in enhancing my work.

File: llms/mlx_lm/tuner/dora.py (hunk 5 above)
Comment (awni) [reply llms/mlx_lm/tuner/dora.py:107]: Yes but I think the paper is working on a transposed version of the weights. Check the suggested implementation here: https://nbasyl.github.io/DoRA-project-page/

File: llms/mlx_lm/tuner/utils.py (hunk 1 above)
Comment (alexC-nonsense4k) [reply llms/mlx_lm/tuner/utils.py]: remove  `linear_to_dora_layers`  and  pass `use_dora` to `linear_to_lora_layers`

File: llms/mlx_lm/tuner/utils.py (hunk 3 above)
Comment (alexC-nonsense4k) [reply llms/mlx_lm/tuner/utils.py]: remove this function `remove_dora_layers`

File: llms/mlx_lm/tuner/utils.py (hunk 2 above)
Comment (alexC-nonsense4k) [reply llms/mlx_lm/tuner/utils.py]: remove  `apply_dora_layers`  and  pass `use_dora` to `apply_lora_layers`

File: llms/mlx_lm/tuner/dora.py (hunk 5 above)
Comment (alexC-nonsense4k) [reply llms/mlx_lm/tuner/dora.py]: Yes, you are right. Now code has been changed to `self.m = mx.linalg.norm(self.linear.weight, axis=1)` in mlx_lm/tuner/dora.py

File: llms/mlx_lm/tuner/dora.py (hunk 7 above)
Comment (alexC-nonsense4k) [reply llms/mlx_lm/tuner/dora.py]: use `mx.linalg.norm` to calculate norm and use mx.stop_gradient to prevent gradients from flowing through `norm` during backpropagation

`norm = mx.stop_gradient(mx.linalg.norm(adapted, axis=1))`

File: llms/mlx_lm/tuner/dora.py (hunk 9 above)
Comment (alexC-nonsense4k) [reply llms/mlx_lm/tuner/dora.py]: removed unnecessary annotation

File: llms/mlx_lm/tuner/dora.py (hunk 8 above)
Comment (alexC-nonsense4k) [reply llms/mlx_lm/tuner/dora.py]: remove  `input_dims *= 32 // linear.bits` and add `raise ValueError("DoRALinear does not yet support quantized weights.")`

File: llms/mlx_lm/tuner/dora.py (hunk 10 above)
Comment (alexC-nonsense4k) [reply llms/mlx_lm/tuner/dora.py]: removed

File: llms/mlx_lm/tuner/dora.py (hunk 11 above)
Comment (alexC-nonsense4k) [reply llms/mlx_lm/tuner/dora.py]: removed

File: llms/mlx_lm/tuner/dora.py (hunk 12 above)
Comment (alexC-nonsense4k) [reply llms/mlx_lm/tuner/dora.py]: use `mx.linalg.norm` instead of `.square().sum(axis=1,keepdims=True).sqrt()`
```
        fused_linear.weight = weight + lora_b @ lora_a

        fused_norm=mx.linalg.norm(fused_linear.weight,axis=1)
```

File: llms/mlx_lm/tuner/dora.py (hunk 13 above)
Comment (alexC-nonsense4k) [reply llms/mlx_lm/tuner/dora.py]: adjusted the order of calculation

File: llms/mlx_lm/tuner/dora.py (hunk 14 above)
Comment (alexC-nonsense4k) [reply llms/mlx_lm/tuner/dora.py]: removed

File: llms/mlx_lm/tuner/dora.py [hunk 15]
```diff
@@ -0,0 +1,119 @@
...
+    def __call__(self,x) :
+        dtype=self.linear.weight.dtype
+        lb = (self.scale * self.lora_b.T).astype(dtype)
+        la = self.lora_a.T.astype(dtype)
+        adapted=self.linear.weight+lb.astype(dtype) @ la
+        norm= adapted.square().sum(axis=0,keepdims=True).sqrt()
+        y = self.linear(x)
+        z = (self.dropout(x) @ self.lora_a) @ self.lora_b
+        res=y+(self.scale*z).astype(x.dtype)
+        return res/norm*self.m
```
[the remark this answers is not in the document]
Comment (alexC-nonsense4k) [reply llms/mlx_lm/tuner/dora.py]: adjusted the order of calculation

**Comment by alexC-nonsense4k** [comment]
I fixed all the problem that mentioned above. Plus I find in mlx_lm/fuse.py and mlx_lm/generate.py, there are functions that related to applying lora or dora layers. I add use_dora in these two files to support dora.

**Review by awni — APPROVED** [review]
Thanks for the addition!!
